Agilex arm sip svc driver

maheshraotm opened the pull request:
Arm SiP SVC services (aka Silicon Provider Supervisory call) provides a generic interface to send
the SMC/HVC call from kernel running at EL1 to hypervisor/secure monitor
firmware running at EL2/EL3.

SiP SVC service supports multiple clients. The client need to "open" before
sending any SMC/HVC request. And, the client must "close" after completion.
The client can input any SMC/HVC a0~a7 register value according to their application.

Besides, sip_svc service also provide 'sip_svc' shell command for testing and debugging purpose.
the subsystem consists of two parts a svc driver and subsystem which provides the service using the driver.

carlocaione commented on subsys/sip_svc/Kconfig:
```diff
@@ -0,0 +1,87 @@
+module = ARM_SIP_SVC_ARM64
+module-str = arm_sip_svc
+
+config ARM_SIP_SVC_HAS_HVC_CALLS
+	bool "Enables HVC calls to EL2 layer"
+	default n
+	depends on ARM_SIP_SVC_ARM64
+
+config ARM_SIP_SVC_HAS_SMC_CALLS
+	bool "Enables SMC calls to EL3 layer"
+	default n
+	depends on ARM_SIP_SVC_ARM64
```
We should leverage the DT for this, not using Kconfig symbols. See for example `drivers/pm_cpu_ops/pm_cpu_ops_psci.c`. We already ship a `method` DT property that defines whether we are going to use HVC or SMC.

maheshraotm replied:
Hi drivers/pm_cpu_ops/pm_cpu_ops_psci.c is a driver implementation.
To achieve the similar implementation here I would need to add a dts entry and yaml file ,but I am not sure if that is the correct way to go in a subsystem.
Also We had initially tried to upstream this as a driver , but was told  that the implementation is more suited as a subsystem. So we made the necessary changes and submitted it here
https://github.com/zephyrproject-rtos/zephyr/pull/43256#pullrequestreview-896090862.
Can you suggest how to go ahead for this?

povergoing replied:
AFAIK, the sip_svc is designed as an interface right? Generally, the whole thing should consist of 2 parts: 1) the interface part which belongs to subsys, and 2) the implementation part which belongs to the driver, and implements the sip_svc subsys. Doing this is because we expect the sip_svc interface/subsys to have many implementations with different platforms.
So this HVC/SVC calls should be the driver part instead of subsys part.
Not sure if I understand right and if I make it clear?

maheshraotm replied:
I have updated the PR, with implementation for driver subsystem split

povergoing commented on subsys/sip_svc/sip_svc_agilex_mailbox_shell.c:
```diff
@@ -0,0 +1,348 @@
+	}
+
+	*hex_val = 0;
+	for (i = len - 1; i >= 0; i--) {
```
ditto

maheshraotm replied:
changed to strtol

povergoing commented on subsys/sip_svc/sip_svc.c:
```diff
@@ -0,0 +1,898 @@
+
+		if (ret != 0) {
+			sip_svc_id_mgr_delete(ctrl->client_id_pool);
+			k_free(ctrl->trans_id_map);
```
line 834 you are using `sip_svc_id_map_delete(ctrl->trans_id_map);` but here is `k_free(ctrl->trans_id_map);` ?

maheshraotm replied:
fixed

povergoing commented on subsys/sip_svc/sip_svc.c:
```diff
@@ -0,0 +1,898 @@
+	uint32_t c_idx;
+
+	if (!ctrl)
+		return SIP_SVC_ID_INVALID;
```
Again, is that better to use `return -EINVAL;` here?

maheshraotm replied:
fixed to use error number return

povergoing commented on include/zephyr/sip_svc/sip_svc.h:
```diff
@@ -0,0 +1,225 @@
+ * @param ctrl Controller structure which provides Arm SiP services.
+ * @param c_token Client's token
+ *
+ * @return address pointer to the client private data
```
address pointer ..... or NULL?

maheshraotm replied:
fixed

povergoing commented on subsys/sip_svc/sip_svc.c:
```diff
@@ -0,0 +1,898 @@
+
+static uint32_t sip_svc_generate_c_token(void)
+{
+	uint32_t c_token = k_cycle_get_32();
```
Not for sure if it is your case, just a kind reminder. Some platform has a low freq timer, when you continuously call `k_cycle_get_32()`, you might get 2 or more same tokens. If it doesn't matter, just ignore this:)

povergoing reviewed (changes requested):
LGTM overall, only some questions and small issues

maheshraotm commented:
Hi , could the stale label be removed .I will update the PR with the refactored code

maheshraotm commented on subsys/sip_svc/sip_svc_shell.c:
```diff
@@ -0,0 +1,393 @@
+	}
+
+	*hex_val = 0;
+	for (i = len - 1; i >= 0; i--) {
```
done

maheshraotm commented on subsys/sip_svc/sip_svc.c:
```diff
@@ -0,0 +1,898 @@
+	uint32_t c_idx = SIP_SVC_ID_INVALID;
+
+	if (!ctrl)
+		return SIP_SVC_ID_INVALID;
```
done

povergoing reviewed (commented):
LGTM, some nits: should the copyright be 2023?

carlocaione commented:
@maheshraotm I really would like to see how this looks from the user prospective. Can you add a test or sample to exercise this subsystem?

maheshraotm commented:
> @maheshraotm I really would like to see how this looks from the user prospective. Can you add a test or sample to exercise this subsystem?

for testing and as a sample we have added two shell based command interface sip_svc_agilex_mailbox_shell.c and sip_svc_shell.c. Do you want something something as a sample application apart from these?

pdgendt commented on subsys/sip_svc/sip_svc.c:
```diff
@@ -0,0 +1,934 @@
+			return ret;
+		}
+
+		/* Create and run the thread */
+		ctrl->tid = k_thread_create(&ctrl->thread, ctrl->stack,
+					CONFIG_ARM_SIP_SVC_THREAD_STACK_SIZE,
+					sip_svc_thread, ctrl, NULL, NULL,
+					CONFIG_ARM_SIP_SVC_THREAD_PRIORITY,
+					K_ESSENTIAL, K_NO_WAIT);
```
General remark:
This file does A LOT of mutex (un)locking, thread suspending/resuming, message queue put/get, ... I wonder if this subsystem logic would work better using a [work queue](https://docs.zephyrproject.org/3.2.0/kernel/services/threads/workqueue.html)?

pdgendt commented on subsys/sip_svc/sip_svc_shell.c:
```diff
@@ -0,0 +1,309 @@
+	if (err < 0)
+		return err;
+
+	priv = (struct private_data *)k_malloc(sizeof(struct private_data));
```
Why use the heap? Put it on the stack and pass a pointer to the callback function?

maheshraotm commented:
> This pull request has been marked as stale because it has been open (more than) 60 days with no activity. Remove the stale label or add a comment saying that you would like to have the label removed otherwise this pull request will automatically be closed in 14 days. Note, that you can always re-open a closed pull request at any time.

I have updated the PR, could you remove the stale tag?

povergoing commented on samples/subsys/sip_svc/README.rst:
```diff
@@ -0,0 +1,31 @@
+********************
+For building the application
+.. zephyr-app-commands::
+    :zephyr-app: samples/subsys/sip_Svc
```
CI warning:
```
Warning, treated as error:
/home/runner/work/zephyr/zephyr/doc/_build/src/samples/subsys/sip_svc/README.rst:25:Unexpected indentation.
FAILED: CMakeFiles/html 
```
https://github.com/zephyrproject-rtos/zephyr/actions/runs/4433960095/jobs/7780005949?pr=50383

povergoing commented on subsys/sip_svc/sip_svc_intel_socfpga.c:
```diff
@@ -0,0 +1,158 @@
+/*
```
I don't have a strong opinion on this, but should this file (containing the platform-specific functions) belong to the `driver/svc/`? The `subsys/svc` is designed to contain the most common, no-platform-specific, and abstracted interface, right? What's your opinion?

maheshraotm replied:
I have made the change to bring the platform specific implementation to the driver side , can you please check if this is ok?

povergoing replied:
Yes, LGTM

maheshraotm committed:
dts: intel: Add dtsi entry for SiP SMC call.

Add SiP SVC driver dtsi entry for smc call in INTEL AGILEX SOC FPGA.

Signed-off-by: Mahesh Rao <mahesh.rao@intel.com>

maheshraotm committed:
samples: sip_svc: Sample application using sip_svc service.

Sample application to retrieve voltage from SDM in
INTEL AGILEX SOC FPGA

Signed-off-by: Mahesh Rao <mahesh.rao@intel.com>

maheshraotm committed:
CODEOWNER: Add code owner for sip_svc subsystem

Add code owner for svc driver,sip_svc subsystem and
sip_svc sample application

Signed-off-by: Mahesh Rao <mahesh.rao@intel.com>

carlocaione commented on include/zephyr/drivers/sip_svc/sip_svc_driver.h:
```diff
@@ -0,0 +1,46 @@
+					 struct arm_smccc_res *res);
+};
+
+struct svc_api {
+	void (*svc_func)(const struct device *dev, unsigned long function_id,
+					 unsigned long arg0, unsigned long arg1, unsigned long arg2,
+					 unsigned long arg3, unsigned long arg4, unsigned long arg5,
+					 unsigned long arg6, struct arm_smccc_res *res);
+	bool (*sip_svc_plat_func_id_valid)(uint32_t command, uint32_t func_id);
+	uint32_t (*sip_svc_plat_format_trans_id)(uint32_t client_idx,
+					 uint32_t trans_idx);
+	uint32_t (*sip_svc_plat_get_trans_idx)(uint32_t trans_id);
+	void (*sip_svc_plat_update_trans_id)(struct sip_svc_request *request,
+					 uint32_t trans_id);
+	void (*sip_svc_plat_free_async_memory)(struct sip_svc_request *request);
+	int (*sip_svc_plat_async_res_req)(unsigned long *a0, unsigned long *a1,
+					 unsigned long *a2, unsigned long *a3, unsigned long *a4,
+					 unsigned long *a5, unsigned long *a6, unsigned long *a7,
+					 char *buf, size_t size);
+	int (*sip_svc_plat_async_res_res)(struct arm_smccc_res *res, char *buf,
+					 size_t *size, uint32_t *trans_id);
+	uint32_t (*sip_svc_plat_get_error_code)(struct arm_smccc_res *res);
+};
```
This is a really odd driver that is basically not following any of the standard (zephyr) way of defining and using drivers. There is a reason for that?

Take for example (this is valid for any zephyr driver subsystem really) the CAN driver subsystem. You have a `__subsystem` marked set of hooks in `include/zephyr/drivers/can.h` that are specialized by a driver without resorting to `__weak` symbols like it is done in your implementation (and without any "default / empty" implementation for that).

maheshraotm replied:
thanks for pointing this out, I will rework the current implementation to this

maheshraotm replied:
I have reworked the implementation with the suggested changes. Could you please check them?

maheshraotm committed:
subsystem: sip_svc: Add ARM SiP SVC subsystem

Introduce a new SiP SVC subsystem to provide ARM Silicon Provider based
supervisory call services. SiP SVC service provides the capability to
send SMC/HVC call from kernel running at EL1 to hypervisor/
secure monitor firmware running at EL2/EL3 and also added SiP SVC shell
commands to exercise the service.

Signed-off-by: Mahesh Rao <mahesh.rao@intel.com>

maheshraotm committed:
subsystem: sip_svc: Add SDM mailbox specific shell

Add shell based on sip_svc subsystem to access the SDM mailbox in
EL3 layer using ASYNC sip_svc request.

Signed-off-by: Mahesh Rao <mahesh.rao@intel.com>

maheshraotm committed:
drivers: sip_svc: Add driver for SiP Supervisory call

Add driver for communicating with EL3/EL2 layer using smc/hvc call
for Silicon vendor Provided services for INTEL AGILEX SOC FPGA.

Signed-off-by: Mahesh Rao <mahesh.rao@intel.com>

carlocaione commented on drivers/sip_svc/sip_svc_intel_socfpga.c:
```diff
@@ -0,0 +1,252 @@
+#include <string.h>
+#include <zephyr/drivers/sip_svc/sip_svc_agilex_mailbox.h>
+#include <zephyr/drivers/sip_svc/sip_svc_agilex_smc.h>
+#include <zephyr/drivers/sip_svc/sip_svc_driver.h>
```
the `sip_svc_driver.h` header should go into `include/zephyr/drivers`.

carlocaione replied:
also it should be called `sip_svc.h`

maheshraotm replied:
There is already a sip_svc.h file for the subsystem public API's. The users are only required to use the public subsystem API's. The sip_svc_driver.h contains the API format for platform specific implementation for sip_svc . So I can move it outside of include/zephyr/drivers/sip_svc/, But this can be renamed to **sip_svc_plat.h**?Is this fine?

carlocaione reviewed (commented):
@maheshraotm do you think you could introduce this driver and subsystem to the next dev review meeting?

MaureenHelm commented:
> @maheshraotm do you think you could introduce this driver and subsystem to the next dev review meeting?

Added to the agenda for next week (Apr 6)

galak reviewed (dismissed):
Seems like devicetree should be `intel,socfpga-agilex-sip-svc`, not `arm,sip_svc`.

carlescufi commented:
Architecture WG:

Presentation by @maheshraotm.

- @carlocaione asks why the split driver/subsystem. He also raises the similarity with the PSCI code
- @maheshraotm states that there are differences between PSCI and communication with the SDM manager
- @bjarki-trackunit claims that the driver API seems to be extremely generic. @maheshraotm clarifies that this is only for the supervisory call API. 
- @carlocaione states that there will be duplication in Devicetree between PSCI and this
- @galak asks if there is a standard. SMC messages are standard, as per @carlocaione.

carlocaione reviewed (dismissed):
there are some nits but nothing preventing this to be merged and this was already blocked long enough.

povergoing reviewed (dismissed):
LGTM

maheshraotm commented:
> Seems like devicetree should be `intel,socfpga-agilex-sip-svc`, not `arm,sip_svc`.

Hi @galak , I have made the required changes now the driver for sip_svc on agilex for smc calles will depend on "intel,agilex-socfpga-sip-smc"

maheshraotm commented:
@galak , @nashif , @carlocaione and @povergoing , I have updated the driver implementation specific to EL3 communication for agilex platform according to the review comments and also added more comments in the code. Can you please review the PR?.
